### ISSUE TITLE
Error while evaluating tflite model with bazel run_eval 

### ISSUE THREAD

**Issue by gabrielecorsi97** [tensorflow/tensorflow#59239]
Hi, i managed to train a SSD_mobilenet_v1 using model_main_tf2.py. After that i managed to export with export_tflite_graph_tf2.py changing the max_detections paramenter to 33, and then i successfully exported a tflite model. I tried then to run some inference with the new tflite model and it worked as expected.
I decided then to try to evaluate  the tflite model using this [guide](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/tools/evaluation/tasks/coco_object_detection),  but when i run this command: 
`bazel run -c opt -- //tensorflow/lite/tools/evaluation/tasks/coco_object_detection:run_eval --model_file=/home/gab/PycharmProjects/tensorflow_prova2/new_model_quant_F16Q.tflite --ground_truth_images_path=/home/gab/PycharmProjects/tensorflow_prova2/images  --model_output_labels=/home/gab/PycharmProjects/tensorflow_prova2/label_map.txt --output_file_path=/home/gab/PycharmProjects/tensorflow_prova2/coco_output.txt --debug_mode=true`
output shows odd results like this:
Object [0]
  Score: 33
  Class-ID: 1
  Bounding Box:
    Normalized Top: 0.999825
    Normalized Bottom: 0.999262
    Normalized Left: 0.999445
    Normalized Right: 0.999066
Object [1]
  Score: 0
  Class-ID: 1
  Bounding Box:
    Normalized Top: 0.998896
    Normalized Bottom: 0.997808
    Normalized Left: 0.998822
    Normalized Right: 0.996969
Object [2]
  Score: 0.780337
  Class-ID: 1
  Bounding Box:
    Normalized Top: 0.996294
    Normalized Bottom: 0.988497
    Normalized Left: 0.994716
    Normalized Right: 0.985629
Object [3]
  Score: 0.479028
  Class-ID: 1
  Bounding Box:
    Normalized Top: 0.972919
    Normalized Bottom: 0.0584207
    Normalized Left: 0.0640493
    Normalized Right: 0.0434026
Object [4]
  Score: 0
  Class-ID: 1
  Bounding Box:
    Normalized Top: 0.0374295
    Normalized Bottom: 0.0290796
    Normalized Left: 0.0313877
    Normalized Right: 0.0227272
Object [5]
  Score: 0
  Class-ID: 1
  Bounding Box:
    Normalized Top: 0.0221782
    Normalized Bottom: 0.0208216
    Normalized Left: 0.0213712
    Normalized Right: 0.0207996
Object [6]
  Score: 0
  Class-ID: 1
  Bounding Box:
    Normalized Top: 0.0199532
    Normalized Bottom: 0.0192215
    Normalized Left: 0.0193954
    Normalized Right: 0.0187497
Object [7]
  Score: 0
  Class-ID: 1
  Bounding Box:
    Normalized Top: 0.0186509
    Normalized Bottom: 0.0180727
    Normalized Left: 0.0181409
    Normalized Right: 0.0176707
Object [8]
  Score: 0
  Class-ID: 1
  Bounding Box:
    Normalized Top: 0.0176005
    Normalized Bottom: 2.02493
    Normalized Left: 0.0637459
    Normalized Right: 0.325256

This is for the first image of the folder i used, one thing i noted is that the first score of the first object is always 33 that is the parameter i changed when i exported the trained model with export_tflite_graph_tf2.py

Is there a way to fix this so i can evaluate my tflite model?
Thanks

**Comment by synandi** [comment]
@gabrielecorsi97 
Could you please provide reproducible code or Colab gist for reproducing the issue reported here?
Thank you!

**Comment by gabrielecorsi97** [comment]
You can download my model [here](https://drive.google.com/file/d/1Wgnhj34Y90kVhEAzfTa48MCEFF5j36RX/view?usp=sharing).
Also my folder images is [here](https://drive.google.com/drive/folders/1M628HsaBgz0BeMpE_9M_CWddtbe7Vy6v?usp=share_link) and my [label_map.txt](https://github.com/tensorflow/tensorflow/files/10401052/label_map.txt)

These are the inputs for running the command from which i get bad results
Tell me if you need all the code from the training to the conversion of the model.
Thanks again.

**Comment by synandi** [comment]
@gabrielecorsi97 
Kindly provide the complete code from the training to the conversion of the model. Thank you!

**Comment by gabrielecorsi97** [comment]
Basically i followed this [tutorial](https://neptune.ai/blog/how-to-train-your-own-object-detector-using-tensorflow-object-detection-api), i selected the [SSD MobileNet V1 FPN 640x640](http://download.tensorflow.org/models/object_detection/tf2/20200711/ssd_mobilenet_v1_fpn_640x640_coco17_tpu-8.tar.gz) from the Model Zoo
I  modified the [pipeling.config](https://drive.google.com/file/d/1FSMeQKjn4alDVAgfwize0pAu-wYQyEPi/view?usp=share_link) and used this [label_map.pbtxt](https://drive.google.com/file/d/1j2SPy9H_Hw9pWFYrDp9Hy_7MTeSTqMIl/view?usp=share_link)
Then i runned this command:
`python model_main_tf2.py
  --pipeline_config_path=<path to your config file>
  --model_dir=<path to a directory with your model>
  --checkpoint_every_n=500
  --num_workers=8
  --alsologtostderr`

And this command:
`python object_detection/export_tflite_graph_tf2.py 
    --pipeline_config_path path/to/ssd_model/pipeline.config 
    --trained_checkpoint_dir path/to/ssd_model/checkpoint 
    --output_directory path/to/exported_model_directory`
but in the file export_tflite_graph_tf2.py i changed the max_detections from 10 to 33

After that i wrote this [script](https://drive.google.com/file/d/1FU95EyCYJZ9U1E3fUp8GogG0KKyEI0YM/view?usp=share_link) to convert the savedModel to 3 tflite model with 3 different quantization method [[model1](https://drive.google.com/file/d/1ob14KHaaBQpP_0ex48X1CzfNqKa7aoZj/view?usp=share_link)][[model2](https://drive.google.com/file/d/1Wgnhj34Y90kVhEAzfTa48MCEFF5j36RX/view?usp=share_link)][[model3](https://drive.google.com/file/d/17xat4qRzvfhZJWEeXoMaOcnnqZoWvjNJ/view?usp=share_link)].
I tried to do some inference with those and all seemed to work as intended (performance seems a bit slow but that is another problem)
At this point i tried to run the run_eval command with bazel but outputs are wrong.

**Comment by gabrielecorsi97** [comment]
I think i found a way to make the run_eval work for me. In the file object_detection_stage.cc i change the index in the lines 117-120 as follows:
int num_detections = static_cast<int>(*static_cast<float*>(outputs->at(2)));
  float* detected_label_boxes = static_cast<float*>(outputs->at(1));
  float* detected_label_indices = static_cast<float*>(outputs->at(3));
  float* detected_label_probabilities = static_cast<float*>(outputs->at(0));
With this change when i try to evaluate my tflite mobilenet model i get these outputs:
Image: ec5_jpg.rf.a3b15ab0256e453798067e0a347c0328.jpg

Object [0]
  Score: 0.999732
  Class-ID: 10
  Bounding Box:
    Normalized Top: 0.538579
    Normalized Bottom: 0.616121
    Normalized Left: 0.232007
    Normalized Right: 0.291118
Object [1]
  Score: 0.999694
  Class-ID: 10
  Bounding Box:
    Normalized Top: 0.61965
    Normalized Bottom: 0.696531
    Normalized Left: 0.418127
    Normalized Right: 0.475554
Object [2]
  Score: 0.998913
  Class-ID: 4
  Bounding Box:
    Normalized Top: 0.283489
    Normalized Bottom: 0.351773
    Normalized Left: 0.646506
    Normalized Right: 0.698427
Object [3]
  Score: 0.998561
  Class-ID: 10
  Bounding Box:
    Normalized Top: 0.170539
    Normalized Bottom: 0.238299
    Normalized Left: 0.2561
    Normalized Right: 0.307592
Object [4]
  Score: 0.998127
  Class-ID: 10
  Bounding Box:
    Normalized Top: 0.332123
    Normalized Bottom: 0.404013
    Normalized Left: 0.411455
    Normalized Right: 0.460659
Object [5]
  Score: 0.997882
  Class-ID: 4
  Bounding Box:
    Normalized Top: 0.333157
    Normalized Bottom: 0.404034
    Normalized Left: 0.491775
    Normalized Right: 0.540915
Object [6]
  Score: 0.99735
  Class-ID: 4
  Bounding Box:
    Normalized Top: 0.463913
    Normalized Bottom: 0.539379
    Normalized Left: 0.597591
    Normalized Right: 0.650279
Object [7]
  Score: 0.99733
  Class-ID: 8
  Bounding Box:
    Normalized Top: 0.34933
    Normalized Bottom: 0.479003
    Normalized Left: 0.229403
    Normalized Right: 0.305899
Object [8]
  Score: 0.996965
  Class-ID: 4
  Bounding Box:
    Normalized Top: 0.400276
    Normalized Bottom: 0.471837
    Normalized Left: 0.670664
    Normalized Right: 0.721426
Object [9]
  Score: 0.996301
  Class-ID: 4
  Bounding Box:
    Normalized Top: 0.175871
    Normalized Bottom: 0.244033
    Normalized Left: 0.496007
    Normalized Right: 0.541796
Object [10]
  Score: 0.995941
  Class-ID: 12
  Bounding Box:
    Normalized Top: 0.165368
    Normalized Bottom: 0.24047
    Normalized Left: 0.175837
    Normalized Right: 0.23584
Object [11]
  Score: 0.995646
  Class-ID: 2
  Bounding Box:
    Normalized Top: 0.421515
    Normalized Bottom: 0.560066
    Normalized Left: 0.758904
    Normalized Right: 0.849036
Object [12]
  Score: 0.992486
  Class-ID: 3
  Bounding Box:
    Normalized Top: 0.38341
    Normalized Bottom: 0.475804
    Normalized Left: 0.40925
    Normalized Right: 0.473162
Object [13]
  Score: 0.991171
  Class-ID: 11
  Bounding Box:
    Normalized Top: 0.302968
    Normalized Bottom: 0.420778
    Normalized Left: 0.327911
    Normalized Right: 0.394248
Object [14]
  Score: 0.989693
  Class-ID: 9
  Bounding Box:
    Normalized Top: 0.211129
    Normalized Bottom: 0.304191
    Normalized Left: 0.475005
    Normalized Right: 0.536662
Object [15]
  Score: 0.989301
  Class-ID: 7
  Bounding Box:
    Normalized Top: 0.518127
    Normalized Bottom: 0.617244
    Normalized Left: 0.406102
    Normalized Right: 0.471927
Object [16]
  Score: 0.988999
  Class-ID: 10
  Bounding Box:
    Normalized Top: 0.276729
    Normalized Bottom: 0.347333
    Normalized Left: 0.26124
    Normalized Right: 0.311395
Object [17]
  Score: 0.986081
  Class-ID: 12
  Bounding Box:
    Normalized Top: 0.616075
    Normalized Bottom: 0.709131
    Normalized Left: 0.218458
    Normalized Right: 0.288195
Object [18]
  Score: 0.977899
  Class-ID: 6
  Bounding Box:
    Normalized Top: 0.609654
    Normalized Bottom: 0.703561
    Normalized Left: 0.798571
    Normalized Right: 0.877587
Object [19]
  Score: 0.975307
  Class-ID: 5
  Bounding Box:
    Normalized Top: 0.300313
    Normalized Bottom: 0.420017
    Normalized Left: 0.565314
    Normalized Right: 0.629695
Object [20]
  Score: 0.962002
  Class-ID: 9
  Bounding Box:
    Normalized Top: 0.515903
    Normalized Bottom: 0.617771
    Normalized Left: 0.125869
    Normalized Right: 0.197614
Object [21]
  Score: 0.0450051
  Class-ID: 12
  Bounding Box:
    Normalized Top: 0.66508
    Normalized Bottom: 0.764215
    Normalized Left: 0.877319
    Normalized Right: 0.961008
Object [22]
  Score: 0.0446676
  Class-ID: 8
  Bounding Box:
    Normalized Top: 0.823962
    Normalized Bottom: 1.00381
    Normalized Left: -0.00830653
    Normalized Right: 0.0553037
Object [23]
  Score: 0.0379813
  Class-ID: 10
  Bounding Box:
    Normalized Top: 0.83234
    Normalized Bottom: 0.981694
    Normalized Left: 0.954265
    Normalized Right: 1.00721
Object [24]
  Score: 0.0316117
  Class-ID: 9
  Bounding Box:
    Normalized Top: 0.680946
    Normalized Bottom: 0.786139
    Normalized Left: 0.0318502
    Normalized Right: 0.10451
Object [25]
  Score: 0.0307972
  Class-ID: 4
  Bounding Box:
    Normalized Top: 0.287242
    Normalized Bottom: 0.349748
    Normalized Left: 0.626914
    Normalized Right: 0.689233

....

I'm not sure if displaying all the object detected even if their score is very low is the expected behavior but final output with the precision seems correct: 
Num evaluation runs: 53
Preprocessing latency: avg=7866.28(us), std_dev=0(us)
Inference latency: avg=183770(us), std_dev=3616(us)
Average Precision [IOU Threshold=0.5]: 0.916775
Average Precision [IOU Threshold=0.55]: 0.915203
Average Precision [IOU Threshold=0.6]: 0.913915
Average Precision [IOU Threshold=0.65]: 0.912595
Average Precision [IOU Threshold=0.7]: 0.906381
Average Precision [IOU Threshold=0.75]: 0.859288
Average Precision [IOU Threshold=0.8]: 0.801814
Average Precision [IOU Threshold=0.85]: 0.648213
Average Precision [IOU Threshold=0.9]: 0.298668
Average Precision [IOU Threshold=0.95]: 0.0243592
Overall mAP: 0.719721

I also tried with a resnet and a centernet but these two don't work as expected. I think the reason is because they expect the input image with different values, right?

**Comment by mohantym** [comment]
Ok @gabrielecorsi97 !
Thanks for the update . 

@sachinprasadhs !
Could you look at this issue.

Thank you!

**Comment by sachinprasadhs** [comment]
What do you mean by results are not as expected for resnet and centernet. 
Basically, all these models use the same preprocessing pipeline for the input image. 
If the results are varying from one model to another, it could be due to TFLite conversion, you may have to check your conversion pipeline, please check the documentation [here](https://www.tensorflow.org/lite/models/convert) for more details. Thanks!
